Baremetal/UPI-SriRamanujam.md

```diff
@@ -0,0 +1,92 @@
+
+## Hardware
+
+* 3 hyper-converged hypervisors
+    * Ryzen 5 3600
+    * 64 GiB RAM
+    * 3x 4TiB HDD
+    * 2x 500GiB SSD in RAID1
+    * 1x 256GiB NVME for boot disk
+
+* 1 NUC I had laying around gathering dust
+    * Intel Core i3-5010U
+    * 16 GiB RAM
+    * 500GiB SSD
+
+## Main cluster
+
+My hypervisors each host an identical workload. The total size of this cluster is 3 control plane nodes, and 9 worker nodes.
+So it splits very nicely three ways. Each hypervisor hosts 1 control plane VM and 3 worker VMs.
+
+* 3 control plane nodes
+    * 4x CPU
+    * 10 GiB RAM
+    * 50 GiB disk
+
+* 9 worker nodes
+    * 8x CPU
+    * 16 GiB RAM
+    * 50 GiB root disk
+    * 4 TiB HDD for workload use
+
+* 1 bootstrap node (temporary, taken down after inital setup is complete)
+    * 4 vCPU
+    * 8 GiB RAM
+    * 120 GiB root disk
+
+## Supporting infrastructure
+
+### Networking
+
+OKD, and especially baremetal UPI OKD, requires a very specific network setup. You will most likely need something more
+flexible than your ISP's router to get everything fully configured. [The documentation](https://docs.okd.io/latest/installing/installing_bare_metal/installing-bare-metal.html#installation-network-user-infra_installing-bare-metal) is very clear on the various DNS records and DHCP static allocations you will need to make, so I won't go into them here.
+
+However, there are a couple extra things that you may want to set for best results. In particular, I make sure that I have [PTR records](https://www.cloudflare.com/learning/dns/dns-records/dns-ptr-record/) set up for all my cluster nodes. This is extremely important as the nodes need a correct PTR record set up for them to auto-discover their hostname. Clusters typically do not set themselves up properly if there are hostname collisions!
+
+
+### API load balancer
+
+I run a separate smaller VM on the NUC as a single-purpose load balancer appliance, running HAProxy.
+
+* 1 load balancer VM
+    * 2x vCPU
+    * 256MiB RAM
+    * 10GiB disk
+
+The HAProxy config is straightforward. I adapted mine from the example config file created by the [ocp4-helpernode](https://github.com/RedHatOfficial/ocp4-helpernode/blob/master/templates/haproxy.cfg.j2) playbook.
+
+## Deployment
+
+I create the VMs on the hypervisors using Terraform. The [Terraform Libvirt provider](https://github.com/dmacvicar/terraform-provider-libvirt) is very, very cool. It's also used by `openshift-install` for its Libvirt-based deployments, so it supports everything needed to deploy OKD nodes. Most importantly, I can use Terraform to supply the VMs with their Ignition configs, which means I don't have to worry about passing kernel args manually or setting up a PXE server to get things going like the official OKD docs would have you do. Terraform also makes it easy to tear down the cluster and reset in case something goes wrong.
```

Comment:
Any particular reason not to use libvirt IPI?

Reply:
I haven't looked into Libvirt IPI very much, admittedly, but it seems as though I don't have the ability to add extra disks to the created VMs nor is it clear if the IPI deployment can handle deploying a cluster across multiple hypervisors. From the documentation I found it seemed geared towards spinning up a couple VMs locally for quick dev tests.

Besides, this gave me a opportunity to have some fun pushing the limits of what Terraform is capable of. 😁

Reply:
> I don't have the ability to add extra disks to the created VMs

Right, I think it can be fixed, once libvirt Machine API gets this feature.

> nor is it clear if the IPI deployment can handle deploying a cluster across multiple hypervisors.

I see, well, that's unlikely to happen (its basically Openstack :))

Reply:
>> nor is it clear if the IPI deployment can handle deploying a cluster across multiple hypervisors.

> I see, well, that's unlikely to happen (its basically Openstack :))

Exactly, haha. I actually seriously investigated deploying Openstack to these boxes so that I could then deploy OKD on top of it, but that was one too many levels of overkill, even for me.

Reply:
> Exactly, haha. I actually seriously investigated deploying Openstack to these boxes so that I could then deploy OKD on top of it, but that was one too many levels of overkill, even for me.

as someone currently working through the _shiftstack_ experience, yes.. "overkill" is a nice word XD